scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulerFrontend.java

```diff
@@ -433,28 +426,89 @@ public JobId submit(Job userJob)
                 throw new SubmissionClosedException(msg);
             }
 
-            UserIdentificationImpl ident = frontendState.checkPermission("submit",
-                                                                         YOU_DO_NOT_HAVE_PERMISSION_TO_SUBMIT_A_JOB);
+            final UserIdentificationImpl ident = frontendState.checkPermission("submit",
+                    YOU_DO_NOT_HAVE_PERMISSION_TO_SUBMIT_A_JOB);
 
-            InternalJob job = frontendState.createJob(userJob, ident);
+            final InternalJob job = frontendState.createJob(userJob, ident);
 
-            schedulingService.submitJob(job);
-
-            frontendState.jobSubmitted(job, ident);
-            return job.getId();
+            final Future<?> submitting = submitPool.submit(() -> {
+                try {
+                    schedulingService.submitJob(job);
+                    frontendState.jobSubmitted(job, ident);
+                } catch (Exception e) {
+                    if (job.getId().longValue() == InternalJob.UNINITIALIZED_JOB_ID) {
+                        logger.error(String.format("Error while submitting job %s. JobID wasn't available yet.", userJob.getName()), e);
+                    } else {
+                        logger.error(String.format("Error while submiting a job[id:%s] ", job.getId().value()), e);
+                    }
+                    throw new RuntimeException(e);
+                } finally {
+                    unregisterSumbittedJob(job.getId());
+                }
+            });
+            return blockUntilJobIdInitialized(userJob, job, submitting);
         } catch (Exception e) {
             logger.warn("Error when submitting job.", e);
             throw e;
         }
     }
 
+    /**
+     * This method blocks until we got job id, or exception occurred.
+     *
+     * if future is done but it got an exception -> throw exception
+     * if future is done but no exception occurred
+     * if id is not initialized throw exception, it means we have some severe bug
+     * if id is initialized it is fine we return id
+     * if future is not done
+     * if id is initialized -> register as waiting and return id
+     * if id is not initialize we continue sleep
+     */
+    private JobId blockUntilJobIdInitialized(Job userJob, InternalJob job, Future<?> submitting) throws JobCreationException {
+        try {
+            while (true) {
+                if (submitting.isDone()) {
+                    try {
+                        submitting.get();
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Extract this nested try block into a separate method. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1141)